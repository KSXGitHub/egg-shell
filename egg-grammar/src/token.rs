mod content;
mod indent;

pub use content::*;
pub use indent::*;
