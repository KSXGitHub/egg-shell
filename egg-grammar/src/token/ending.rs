mod comment;
pub mod embed;

pub use comment::CommentToken;
pub use embed::{DocToken, TextToken};

use derive_more::{From, TryInto};

/// Token at the end of a line.
#[derive(Debug, Clone, PartialEq, Eq, From, TryInto)]
#[non_exhaustive]
pub enum EndingToken<Content> {
    /// Single line comment.
    ///
    /// A single line comment starts with the character `#`.
    Comment(CommentToken<Content>),

    /// Multi-line string or embedded code.
    String(TextToken<Content>),

    /// Documentation annotation.
    Doc(DocToken<Content>),
}
