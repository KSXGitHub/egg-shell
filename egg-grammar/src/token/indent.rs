mod indent_char;

pub use indent_char::*;

pub struct IndentToken(Vec<IndentChar>);
