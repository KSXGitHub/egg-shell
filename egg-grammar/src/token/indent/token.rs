use super::IndentChar;
use derive_more::{AsMut, AsRef, Deref, DerefMut, From, Into, IntoIterator};
use itertools::Itertools;
use split_first_char::split_first_char;
use std::fmt::{self, Debug, Display, Formatter};

/// Token of indentation.
#[derive(Clone, PartialEq, Eq, AsMut, AsRef, Deref, DerefMut, From, Into, IntoIterator)]
pub struct IndentToken(Vec<IndentChar>);

impl IndentToken {
    /// Parse a line of text into a pair of indentation and remaining string.
    ///
    /// **Notes:**
    /// * `line` is assumed to not contain any EOL characters.
    pub fn parse(mut line: &str) -> (Self, &'_ str) {
        let mut indent_char_list = Vec::with_capacity(line.len());
        while let Some((first, rest)) = split_first_char(line) {
            let Ok(indent) = first.try_into() else {
                break;
            };
            indent_char_list.push(indent);
            line = rest;
        }
        indent_char_list.shrink_to_fit();
        (IndentToken(indent_char_list), line)
    }

    /// Check if the indent is the start of another indent.
    pub fn is_start_of(&self, other: &[IndentChar]) -> bool {
        other.starts_with(self)
    }

    /// Check if the indent is the shorter start of another indent.
    pub fn is_shorter_start_of(&self, other: &[IndentChar]) -> bool {
        self.len() > other.len() && self.is_start_of(other)
    }
}

impl Display for IndentToken {
    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
        for char in self.iter() {
            write!(f, "{}", char)?
        }
        Ok(())
    }
}

impl Debug for IndentToken {
    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
        write!(f, "IndentToken [")?;

        let consecutive_indents =
            self.iter()
                .map(|&indent| (indent, 1u32))
                .coalesce(|(a, ac), (b, bc)| {
                    if a == b {
                        Ok((a, ac + bc))
                    } else {
                        Err(((a, ac), (b, bc)))
                    }
                });

        for (indent, count) in consecutive_indents {
            let abbr = indent.abbr();
            if count > 1 {
                write!(f, "<{abbr}✕{count}>")?;
            } else {
                write!(f, "<{abbr}>")?;
            }
        }

        write!(f, "]")
    }
}

#[cfg(test)]
mod test {
    use super::*;
    use pretty_assertions::assert_eq;

    #[test]
    fn parse_line() {
        macro_rules! test_case {
            ($input:literal -> [$($indent:ident),* $(,)?], $rest:literal) => {{
                eprintln!("TEST: {:?}", $input);
                assert_eq!(
                    IndentToken::parse($input),
                    (IndentToken(vec![$(IndentChar::$indent),*]), $rest),
                );
            }};
        }

        test_case!("" -> [], "");
        test_case!("abc" -> [], "abc");
        test_case!("\tabc" -> [Tab], "abc");
        test_case!("\t\tabc" -> [Tab, Tab], "abc");
        test_case!("  abc" -> [Space, Space], "abc");
        test_case!("    abc" -> [Space, Space, Space, Space], "abc");
        test_case!(" \t \tabc" -> [Space, Tab, Space, Tab], "abc");
        test_case!("  abc def ghi" -> [Space, Space], "abc def ghi");
        test_case!("\tabc def\tghi" -> [Tab], "abc def\tghi");
    }

    #[test]
    fn display_fmt() {
        macro_rules! str_fmt {
            ($($name:ident),* $(,)?) => {
                IndentToken(vec![$(IndentChar::$name),*]).to_string()
            };
        }

        assert_eq!(str_fmt!(), "");
        assert_eq!(str_fmt!(Space), " ");
        assert_eq!(str_fmt!(Tab), "\t");
        assert_eq!(str_fmt!(Space, Space, Space, Space), " ".repeat(4),);
        assert_eq!(
            str_fmt!(Space, Space, Tab, Tab, Tab, Space),
            format!("{spc}{spc}{tab}{tab}{tab}{spc}", spc = " ", tab = "\t"),
        );
    }

    #[test]
    fn debug_fmt() {
        macro_rules! dbg_fmt {
            ($($name:ident),* $(,)?) => {{
                let indent_token = IndentToken(vec![$(IndentChar::$name),*]);
                format!("{indent_token:?}")
            }};
        }

        assert_eq!(dbg_fmt!(), "IndentToken []");
        assert_eq!(dbg_fmt!(Space), "IndentToken [<SPC>]");
        assert_eq!(dbg_fmt!(Tab), "IndentToken [<TAB>]");
        assert_eq!(
            dbg_fmt!(Space, Space, Space, Space),
            "IndentToken [<SPC✕4>]",
        );
        assert_eq!(
            dbg_fmt!(Space, Space, Tab, Tab, Tab, Space),
            "IndentToken [<SPC✕2><TAB✕3><SPC>]",
        );
    }
}
