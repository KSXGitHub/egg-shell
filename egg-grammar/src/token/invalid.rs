use derive_more::Error;
use split_first_char::split_first_char;
use std::fmt::{self, Display, Formatter};

/// Token represents an invalid (unparsable) character.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Error)]
pub struct InvalidToken(#[error(not(source))] pub char);

impl InvalidToken {
    /// Create an [`InvalidToken`] from the first character of a string.
    pub fn parse(input: &str) -> Option<(Self, &'_ str)> {
        let (char, rest) = split_first_char(input)?;
        let token = InvalidToken(char);
        Some((token, rest))
    }
}

impl Display for InvalidToken {
    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
        let InvalidToken(char) = self;
        let code = *char as u32;
        write!(f, "Unexpected token {char:?} (U+{code:04X})")
    }
}
