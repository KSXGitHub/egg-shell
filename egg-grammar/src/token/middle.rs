pub mod bracket;
pub mod number;
mod operator;
pub mod string;
mod token;
pub mod word;

pub use bracket::BracketToken;
pub use number::NumberToken;
pub use operator::OperatorToken;
pub use string::StringToken;
pub use token::MiddleToken;
pub use word::WordToken;
