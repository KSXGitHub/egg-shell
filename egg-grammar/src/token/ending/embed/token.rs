use super::EmbedTokenBuilder;
use crate::token::{
    IndentToken, InsertWhitespaces, ParseEmbedTokenAttr, ParseEmbedTokenBody, ParseEmbedTokenTag,
};

/// Token for a chunk of embedded lines.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct EmbedToken<Tag, Attr, Body> {
    pub header: (Tag, Attr),
    pub body: Vec<Body>,
}

impl<'input, Tag, Attr, Body> EmbedToken<Tag, Attr, Body>
where
    Tag: ParseEmbedTokenTag<&'input str>,
    Attr: ParseEmbedTokenAttr<&'input str>,
    Body: ParseEmbedTokenBody<&'input str>,
    Vec<Body>: InsertWhitespaces<&'input str>,
{
    /// Build an [`EmbedToken`] from start to finish.
    pub(crate) fn build<'header_indent>(
        header_indent: &'header_indent IndentToken,
        header_text: &'input str,
        next_line: impl FnMut() -> Option<&'input str>,
    ) -> Option<Self> {
        EmbedTokenBuilder::build(header_indent, header_text, next_line)
    }
}

impl<'a, Tag, Attr, Body> EmbedToken<Tag, Attr, Body>
where
    Tag: ParseEmbedTokenTag<&'a str>,
    Attr: ParseEmbedTokenAttr<&'a str>,
{
    /// Parse the header of the embed token.
    pub fn parse_header(input: &'a str) -> Option<(Tag, Attr)> {
        let (tag, input) = Tag::parse(input)?;
        let attr = Attr::parse(input)?;
        Some((tag, attr))
    }

    /// Start the parsing process with a header and an existing body.
    ///
    /// The existing body is typically an empty `Vec` constructed by `Vec::new` or `Vec::with_capacity`.
    pub fn start_parsing(input: &'a str, body: Vec<Body>) -> Option<Self> {
        let header = Self::parse_header(input)?;
        let token = EmbedToken { header, body };
        Some(token)
    }

    /// Start the parsing process with an empty body.
    pub fn new(input: &'a str) -> Option<Self> {
        Self::start_parsing(input, Vec::new())
    }

    /// Start the parsing process with an empty body with a specified capacity.
    pub fn with_capacity(input: &'a str, capacity: usize) -> Option<Self> {
        Self::start_parsing(input, Vec::with_capacity(capacity))
    }
}

impl<'a, Tag, Attr, Body> EmbedToken<Tag, Attr, Body>
where
    Body: ParseEmbedTokenBody<&'a str>,
{
    /// Parse an item and add it to the body.
    pub fn parse_body_item(&mut self, input: &'a str) -> Option<()> {
        let item = Body::parse(input)?;
        self.body.push(item);
        Some(())
    }
}

impl<'a, Tag, Attr, Body> EmbedToken<Tag, Attr, Body>
where
    Vec<Body>: InsertWhitespaces<&'a str>,
{
    /// Parse a string of whitespaces and add it to the body.
    pub fn insert_body_ws(&mut self, ws: &'a str) -> Option<()> {
        self.body.insert_whitespaces(ws)
    }
}
