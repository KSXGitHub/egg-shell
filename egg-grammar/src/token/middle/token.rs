use super::{BracketToken, NumberToken, OperatorToken, StringToken, WhitespaceToken, WordToken};
use crate::token::ParseMiddleToken;
use derive_more::{From, TryInto};

/// Token in the middle of the line.
#[derive(Debug, Clone, Copy, PartialEq, Eq, From, TryInto)]
#[non_exhaustive]
pub enum MiddleToken<Content> {
    /// Single character of Space or Tab.
    Whitespace(WhitespaceToken),

    /// String, exotic syntax, or abnormal identifier.
    ///
    /// **Including**
    ///
    /// * String: `"abc"`, `'abc'`, `prefix"abc"`, `prefix'abc'`, etc.
    /// * Exotic syntax: `rg"[a-z]+"`, `gl"src/**/*.rs"`, etc.
    /// * Abnormal identifier: `id"abc"`, `id'abc'`, `idl"egg: abc"`, `idl'egg: abc'`, etc.
    ///
    /// **Excluding**
    ///
    /// * Multi-line string.
    String(StringToken<Content>),

    /// Keyword or normal identifier.
    Word(WordToken<Content>),

    /// Number.
    ///
    /// **Including**
    ///
    /// * Positive decimal integer: `123`, `123i32`, `123_456_789`, etc.
    /// * Positive hexadecimal integer: `0x1234ABCD`, `0x1234abcd`, `0x1234ABCD_i32`, etc.
    /// * Positive floating point: `123.45`, `123.45f`, `123.45f32`, etc.
    ///
    /// **Excluding**
    ///
    /// * Negative numbers. Composing `-` and number instead.
    /// * `nan` and `inf`. They are keywords.
    Number(NumberToken<Content>),

    /// Round bracket, square bracket, curly bracket, open or closed.
    Bracket(BracketToken),

    /// A sequence of special characters.
    ///
    /// **Including**
    ///
    /// * Expression operator: `+`, `-`, `*`, `/`, `<`, `<=`, `>`, `>=`, `==`, `!=`, etc.
    /// * Path separator (`::`).
    /// * Dot notation (`.`).
    /// * Type specifier (`:`).
    /// * Binding or assignment (`=`).
    /// * etc.
    Operator(OperatorToken<Content>),
}

impl<'a> ParseMiddleToken<&'a str> for MiddleToken<&'a str> {
    fn parse(input: &'a str) -> Option<(Self, &'a str)> {
        macro_rules! try_parse {
            ($token_type:ident) => {
                if let Some((token, rest)) = $token_type::parse(input) {
                    return Some((MiddleToken::from(token), rest));
                }
            };
        }
        try_parse!(WhitespaceToken);
        try_parse!(StringToken);
        try_parse!(WordToken);
        try_parse!(NumberToken);
        try_parse!(BracketToken);
        try_parse!(OperatorToken);
        None
    }
}

#[cfg(test)]
mod test {
    use super::*;
    use pretty_assertions::assert_eq;

    macro_rules! def_macro {
        ($name:ident -> $token_variant:ident) => {
            /// This macro will not test the exact value of the token, only the top-level variant.
            macro_rules! $name {
                ($input:literal -> $rest:literal) => {{
                    eprintln!();
                    eprintln!("TEST: {:?}", $input);
                    let (token, rest) = MiddleToken::parse($input).expect("Parse failure");
                    dbg!(token);
                    assert!(matches!(token, MiddleToken::$token_variant(_)));
                    dbg!(rest);
                    assert_eq!(rest, $rest);
                }};
            }
        };
    }

    #[test]
    fn whitespace() {
        def_macro!(case -> Whitespace);
        case!(" " -> "");
        case!("\t" -> "");
        case!("  " -> " ");
        case!("\t\t" -> "\t");
        case!(" abc" -> "abc");
        case!("\tabc" -> "abc");
    }

    #[test]
    fn string() {
        def_macro!(case -> String);
        case!("''" -> "");
        case!("'abc def ghi'" -> "");
        case!(r#""abc def ghi""# -> "");
        case!("1foo-bar'abc def'" -> "");
        case!(r"prefix'abc def \' ghi\n\t'suffix++' jkl mno'" -> "++' jkl mno'");
        case!("GalaxyBrainGigachadWarStratagem'三十六計，走為上計'逃げるんだよ" -> "逃げるんだよ");
        case!("'" -> "");
        case!("\"" -> "");
        case!("prefix'abc def ghi" -> "");
    }

    #[test]
    fn word() {
        def_macro!(case -> Word);
        case!("print('hello world')" -> "('hello world')");
        case!("if a + b == c then" -> " a + b == c then");
    }

    #[test]
    fn number() {
        def_macro!(case -> Number);
        case!("123_456-789" -> "-789");
        case!("123.45f64+789" -> "+789");
        case!("123.45.rest" -> ".rest");
        case!("0x123ABCi32 rest" -> " rest");
    }

    #[test]
    fn bracket() {
        def_macro!(case -> Bracket);
        case!("(abc + 123)" -> "abc + 123)");
        case!("[0, 1, 2, 3]" -> "0, 1, 2, 3]");
        case!("{ a: 0, b: 1 }" -> " a: 0, b: 1 }");
        case!(")" -> "");
        case!("]" -> "");
        case!("}" -> "");
    }

    #[test]
    fn operator() {
        def_macro!(case -> Operator);
        case!("+=123" -> "123");
        case!("+#abc" -> "abc");
        case!("+ #abc" -> " #abc");
        case!("$interpolation" -> "interpolation");
        case!("@attribute" -> "attribute");
    }

    #[test]
    fn negative() {
        macro_rules! case {
            ($input:literal) => {{
                eprintln!("TEST: {:?}", $input);
                assert_eq!(MiddleToken::parse($input), None);
            }};
        }

        case!("");
        case!("# this is a comment");
    }
}
