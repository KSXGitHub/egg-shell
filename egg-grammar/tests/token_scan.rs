use egg_grammar::token::{Scan, TokenLine};
use exec_diff::assert_eq_uni_diff;
use itertools::Itertools;
use pretty_assertions::assert_eq;

fn title(title: &str) {
    eprintln!();
    eprintln!("TEST: {}", title);
}

macro_rules! test_snapshot {
    ($tokens:expr, $path:literal) => {{
        title("Snapshot of the Debug format the tokens");
        let received = format!("{:#?}", &$tokens);
        let expected = include_str!($path);
        assert_eq_uni_diff(received.trim(), expected.trim());
    }};
}

fn test_ln_text(tokens: &[TokenLine<&str>], text: &str) {
    title("Equality between all_ln_text and the original source");
    let received: Vec<_> = tokens
        .iter()
        .flat_map(TokenLine::all_ln_text)
        .copied()
        .collect();
    let expected: Vec<_> = text.lines().collect();
    assert_eq!(&received, &expected);
}

fn test_src_text(tokens: &[TokenLine<&str>]) {
    title("Equality between ln_text and sum of src_text");
    for item in tokens {
        let ln_text = dbg!(item.ln_text);
        let indent = dbg!(item.indent.src_text);
        let middle = dbg!(item.middle.iter().map(|item| item.src_text).join(""));
        let ending = dbg!(item.ending.as_ref().map_or("", |item| item.src_text.0));
        let sum = dbg!(format!("{indent}{middle}{ending}"));
        assert_eq!(&sum, ln_text);
    }
}

#[test]
fn hello_world() {
    let text = include_str!("fixtures/hello-world.egg");
    let tokens: Vec<_> = dbg!(Scan::new(text).collect());
    test_snapshot!(tokens, "snapshots/token-scan/hello-world.txt");
    test_ln_text(&tokens, text);
    test_src_text(&tokens);
}

#[test]
fn multi_line() {
    let text = include_str!("fixtures/multi-line.egg");
    let tokens: Vec<_> = dbg!(Scan::new(text).collect());
    test_snapshot!(tokens, "snapshots/token-scan/multi-line.txt");
    test_ln_text(&tokens, text);
    test_src_text(&tokens);
}

#[test]
fn invalid_char_heart() {
    let text = "print 'hello world' with ❤️";
    let tokens: Vec<_> = dbg!(Scan::new(text).collect());
    test_snapshot!(tokens, "snapshots/token-scan/invalid-char-heart.txt");
    test_ln_text(&tokens, text);
    test_src_text(&tokens);
}

#[test]
fn invalid_char_nul() {
    let text = "print 'hello world'\0";
    let tokens: Vec<_> = dbg!(Scan::new(text).collect());
    test_snapshot!(tokens, "snapshots/token-scan/invalid-char-nul.txt");
    test_ln_text(&tokens, text);
    test_src_text(&tokens);
}
