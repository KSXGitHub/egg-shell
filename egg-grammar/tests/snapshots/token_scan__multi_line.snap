---
source: egg-grammar/tests/token_scan.rs
expression: tokens
---
[
    TokenLine {
        ln_text: "@@desc Simple hello world text.",
        indent: TokenLineItem {
            src_text: "",
            token: IndentToken [],
        },
        middle: [],
        ending: Some(
            TokenLineItem {
                src_text: (
                    "@@desc Simple hello world text.",
                    [],
                ),
                token: Doc(
                    EmbedToken {
                        header: (
                            DocTokenTag(
                                Some(
                                    Identifier(
                                        "desc",
                                    ),
                                ),
                            ),
                            RawToken(
                                " Simple hello world text.",
                            ),
                        ),
                        body: [],
                    },
                ),
            },
        ),
    },
    TokenLine {
        ln_text: "pub const HELLO_WORLD: str = '''",
        indent: TokenLineItem {
            src_text: "",
            token: IndentToken [],
        },
        middle: [
            TokenLineItem {
                src_text: "pub",
                token: Ok(
                    Word(
                        Keyword(
                            Pub,
                        ),
                    ),
                ),
            },
            TokenLineItem {
                src_text: " ",
                token: Ok(
                    Whitespace(
                        WhitespaceToken(
                            " ",
                        ),
                    ),
                ),
            },
            TokenLineItem {
                src_text: "const",
                token: Ok(
                    Word(
                        Keyword(
                            Const,
                        ),
                    ),
                ),
            },
            TokenLineItem {
                src_text: " ",
                token: Ok(
                    Whitespace(
                        WhitespaceToken(
                            " ",
                        ),
                    ),
                ),
            },
            TokenLineItem {
                src_text: "HELLO_WORLD",
                token: Ok(
                    Word(
                        Identifier(
                            "HELLO_WORLD",
                        ),
                    ),
                ),
            },
            TokenLineItem {
                src_text: ":",
                token: Ok(
                    Operator(
                        OperatorToken(
                            ":",
                        ),
                    ),
                ),
            },
            TokenLineItem {
                src_text: " ",
                token: Ok(
                    Whitespace(
                        WhitespaceToken(
                            " ",
                        ),
                    ),
                ),
            },
            TokenLineItem {
                src_text: "str",
                token: Ok(
                    Word(
                        Keyword(
                            Str,
                        ),
                    ),
                ),
            },
            TokenLineItem {
                src_text: " ",
                token: Ok(
                    Whitespace(
                        WhitespaceToken(
                            " ",
                        ),
                    ),
                ),
            },
            TokenLineItem {
                src_text: "=",
                token: Ok(
                    Operator(
                        OperatorToken(
                            "=",
                        ),
                    ),
                ),
            },
            TokenLineItem {
                src_text: " ",
                token: Ok(
                    Whitespace(
                        WhitespaceToken(
                            " ",
                        ),
                    ),
                ),
            },
        ],
        ending: Some(
            TokenLineItem {
                src_text: (
                    "'''",
                    [
                        "    Hello, World!",
                        "",
                    ],
                ),
                token: Text(
                    EmbedToken {
                        header: (
                            Single,
                            RawToken(
                                "",
                            ),
                        ),
                        body: [
                            RawToken(
                                "Hello, World!",
                            ),
                            RawToken(
                                "",
                            ),
                        ],
                    },
                ),
            },
        ),
    },
    TokenLine {
        ln_text: "@@desc Example YAML text.",
        indent: TokenLineItem {
            src_text: "",
            token: IndentToken [],
        },
        middle: [],
        ending: Some(
            TokenLineItem {
                src_text: (
                    "@@desc Example YAML text.",
                    [],
                ),
                token: Doc(
                    EmbedToken {
                        header: (
                            DocTokenTag(
                                Some(
                                    Identifier(
                                        "desc",
                                    ),
                                ),
                            ),
                            RawToken(
                                " Example YAML text.",
                            ),
                        ),
                        body: [],
                    },
                ),
            },
        ),
    },
    TokenLine {
        ln_text: "pub const CONFIG: str = '''yaml",
        indent: TokenLineItem {
            src_text: "",
            token: IndentToken [],
        },
        middle: [
            TokenLineItem {
                src_text: "pub",
                token: Ok(
                    Word(
                        Keyword(
                            Pub,
                        ),
                    ),
                ),
            },
            TokenLineItem {
                src_text: " ",
                token: Ok(
                    Whitespace(
                        WhitespaceToken(
                            " ",
                        ),
                    ),
                ),
            },
            TokenLineItem {
                src_text: "const",
                token: Ok(
                    Word(
                        Keyword(
                            Const,
                        ),
                    ),
                ),
            },
            TokenLineItem {
                src_text: " ",
                token: Ok(
                    Whitespace(
                        WhitespaceToken(
                            " ",
                        ),
                    ),
                ),
            },
            TokenLineItem {
                src_text: "CONFIG",
                token: Ok(
                    Word(
                        Identifier(
                            "CONFIG",
                        ),
                    ),
                ),
            },
            TokenLineItem {
                src_text: ":",
                token: Ok(
                    Operator(
                        OperatorToken(
                            ":",
                        ),
                    ),
                ),
            },
            TokenLineItem {
                src_text: " ",
                token: Ok(
                    Whitespace(
                        WhitespaceToken(
                            " ",
                        ),
                    ),
                ),
            },
            TokenLineItem {
                src_text: "str",
                token: Ok(
                    Word(
                        Keyword(
                            Str,
                        ),
                    ),
                ),
            },
            TokenLineItem {
                src_text: " ",
                token: Ok(
                    Whitespace(
                        WhitespaceToken(
                            " ",
                        ),
                    ),
                ),
            },
            TokenLineItem {
                src_text: "=",
                token: Ok(
                    Operator(
                        OperatorToken(
                            "=",
                        ),
                    ),
                ),
            },
            TokenLineItem {
                src_text: " ",
                token: Ok(
                    Whitespace(
                        WhitespaceToken(
                            " ",
                        ),
                    ),
                ),
            },
        ],
        ending: Some(
            TokenLineItem {
                src_text: (
                    "'''yaml",
                    [
                        "    title: This is some YAML",
                        "    publisher: The name of the publisher",
                        "    pages: 335",
                        "    chapters: 12",
                        "    genres:",
                        "        - drama",
                        "        - tragedy",
                        "        - comedy",
                        "    contributors:",
                        "        author: Some guy",
                        "        editor: Some other guy",
                        "",
                    ],
                ),
                token: Text(
                    EmbedToken {
                        header: (
                            Single,
                            RawToken(
                                "yaml",
                            ),
                        ),
                        body: [
                            RawToken(
                                "title: This is some YAML",
                            ),
                            RawToken(
                                "publisher: The name of the publisher",
                            ),
                            RawToken(
                                "pages: 335",
                            ),
                            RawToken(
                                "chapters: 12",
                            ),
                            RawToken(
                                "genres:",
                            ),
                            RawToken(
                                "    - drama",
                            ),
                            RawToken(
                                "    - tragedy",
                            ),
                            RawToken(
                                "    - comedy",
                            ),
                            RawToken(
                                "contributors:",
                            ),
                            RawToken(
                                "    author: Some guy",
                            ),
                            RawToken(
                                "    editor: Some other guy",
                            ),
                            RawToken(
                                "",
                            ),
                        ],
                    },
                ),
            },
        ),
    },
    TokenLine {
        ln_text: "@@desc Example JavaScript text.",
        indent: TokenLineItem {
            src_text: "",
            token: IndentToken [],
        },
        middle: [],
        ending: Some(
            TokenLineItem {
                src_text: (
                    "@@desc Example JavaScript text.",
                    [],
                ),
                token: Doc(
                    EmbedToken {
                        header: (
                            DocTokenTag(
                                Some(
                                    Identifier(
                                        "desc",
                                    ),
                                ),
                            ),
                            RawToken(
                                " Example JavaScript text.",
                            ),
                        ),
                        body: [],
                    },
                ),
            },
        ),
    },
    TokenLine {
        ln_text: "pub const PROGRAM: str = '''js",
        indent: TokenLineItem {
            src_text: "",
            token: IndentToken [],
        },
        middle: [
            TokenLineItem {
                src_text: "pub",
                token: Ok(
                    Word(
                        Keyword(
                            Pub,
                        ),
                    ),
                ),
            },
            TokenLineItem {
                src_text: " ",
                token: Ok(
                    Whitespace(
                        WhitespaceToken(
                            " ",
                        ),
                    ),
                ),
            },
            TokenLineItem {
                src_text: "const",
                token: Ok(
                    Word(
                        Keyword(
                            Const,
                        ),
                    ),
                ),
            },
            TokenLineItem {
                src_text: " ",
                token: Ok(
                    Whitespace(
                        WhitespaceToken(
                            " ",
                        ),
                    ),
                ),
            },
            TokenLineItem {
                src_text: "PROGRAM",
                token: Ok(
                    Word(
                        Identifier(
                            "PROGRAM",
                        ),
                    ),
                ),
            },
            TokenLineItem {
                src_text: ":",
                token: Ok(
                    Operator(
                        OperatorToken(
                            ":",
                        ),
                    ),
                ),
            },
            TokenLineItem {
                src_text: " ",
                token: Ok(
                    Whitespace(
                        WhitespaceToken(
                            " ",
                        ),
                    ),
                ),
            },
            TokenLineItem {
                src_text: "str",
                token: Ok(
                    Word(
                        Keyword(
                            Str,
                        ),
                    ),
                ),
            },
            TokenLineItem {
                src_text: " ",
                token: Ok(
                    Whitespace(
                        WhitespaceToken(
                            " ",
                        ),
                    ),
                ),
            },
            TokenLineItem {
                src_text: "=",
                token: Ok(
                    Operator(
                        OperatorToken(
                            "=",
                        ),
                    ),
                ),
            },
            TokenLineItem {
                src_text: " ",
                token: Ok(
                    Whitespace(
                        WhitespaceToken(
                            " ",
                        ),
                    ),
                ),
            },
        ],
        ending: Some(
            TokenLineItem {
                src_text: (
                    "'''js",
                    [
                        "    export function add(a, b) {",
                        "        return a + b",
                        "    }",
                    ],
                ),
                token: Text(
                    EmbedToken {
                        header: (
                            Single,
                            RawToken(
                                "js",
                            ),
                        ),
                        body: [
                            RawToken(
                                "export function add(a, b) {",
                            ),
                            RawToken(
                                "    return a + b",
                            ),
                            RawToken(
                                "}",
                            ),
                        ],
                    },
                ),
            },
        ),
    },
]
